x/gamm/keeper/msg_server.go

```diff
@@ -84,6 +84,22 @@ func (server msgServer) CreatePool(goCtx context.Context, msg types.CreatePoolMs
 	return poolId, nil
 }
 
+// JoinPool routes `JoinPoolNoSwap` where we do an abstract calculation on needed lp liquidity coins to get the designated
+// amount of share for the pool. (This is done by taking the number of shares we want and then using total number of shares
+// to get the ratio of the pool it accounts for. Using this ratio, we iterate through all pool assets to get how much tokens we need
+// to get the specified number of shares).
+// Using the number of tokens needed to actually join the pool, we do a basic sanity check on whether the token does not exceed
+// `TokenInMaxs`. Then we hit the actual implementation of `JoinPool` defined by each pool model.
+// `JoinPool` takes in the tokensIn calculated above as the parameter rather than using the number of shares provided in the msg.
+// This can result in negotiable difference between the number of shares provided within the msg
+// and the actual number of share amount resulted from joining pool.
+// Internal logic flow for each pool model is as follows:
+// Balancer: TokensIn provided as the argument must be either a single token or tokens containing all assets in the pool.
+// 			 For the case of a single token, we simply perform single asset join (balancer notation: pAo, pool shares amount out,
+// 			 given single asset in).
+//			 For the case of multi-asset join, we first calculate the maximal amount of tokens that can be joined whilst maintaining
+// 			 pool asset's ratio without swap. We then iterate through the remaining coins that couldn't be joined
+// 			 and perform single asset join on each token.
```

Comment:
What do you think about moving this to a README? 

While this is very useful for a doc, I don't think it should be a method spec. For the same reason I mentioned earlier - method/function specs should not be concerned about implementation details to avoid breaking the black box.

I understand that the goal here is to simplify the review process for the community/validators. However, while doing so, we should aim to avoid violating engineering practices

Reply:
So I suggest keeping a short description for the spec with regards to inputs and outputs and moving this excellent write-up to a README

Reply:
I think we make a second README / spec file with the detail, but the overall logic being here seems good

Reply:
I also don't think this violated any engineering principles, this describes how the result is computed, which is something a caller should care about

Nothing sticks out to me as a logic violation

Reply:
If this were a single balancer implementation.

It does stick out as abstraction violation given the pool abstraction

Reply:
I think we should iterate towards a better long term structure!